runner/utils.py

```diff
@@ -123,11 +123,19 @@ def append_parent_path_and_id_(pipeline: Pipeline) -> None:
     ]
 
     for hierarchy in hierarchies:
+        seen_datasources = {}
 
         def append(block, parent_path: str, id_with_prefix: str):
             block.obj.parent_path = parent_path
             if not isinstance(block.obj, DataSource):
                 block.obj.id += id_with_prefix
+            elif mask:
+                if block.obj.id not in seen_datasources.keys():
+                    original_block_obj_id = copy(block.obj.id)
```

Comment:
random thing: you don't need to copy a string, it's immutable, you can't accidentally modify it